crates/biome_js_analyze/src/analyzers/nursery/no_undeclared_dependencies.rs

```diff
@@ -44,16 +44,39 @@ impl Rule for NoUndeclaredDependencies {
 
     fn run(ctx: &RuleContext<Self>) -> Self::Signals {
         let node = ctx.query();
-        let text = node.inner_string_text()?;
-        if !text.text().starts_with('.')
-            // Ignore imports using a protocol such as `node:`, `bun:`, `jsr:`, `https:`, and so on.
-            && !text.text().contains(':')
-            && !ctx.is_dependency(text.text())
-            && !ctx.is_dev_dependency(text.text())
-        {
-            return Some(());
+
+        let token_text = node.inner_string_text()?;
+        let text = token_text.text();
+
+        // Ignore relative path imports
+        // Ignore imports using a protocol such as `node:`, `bun:`, `jsr:`, `https:`, and so on.
+        if text.starts_with('.') || text.contains(':') {
+            return None;
+        }
+
+        let package_name = match text.chars().position(|c| c == '/') {
+            Some(slash_index) => {
+                // scoped package: @mui/material/Button
+                // the package name is @mui/material, not @mui
+                if text.starts_with('@') {
+                    &text[..text[slash_index + 1..]
+                        .chars()
+                        .position(|c| c == '/')
+                        .map_or(text.len(), |i| i + slash_index + 1)]
+                }
+                // unscoped package
+                else {
+                    &text[..slash_index]
+                }
+            }
+            None => text,
+        };
```

Comment:
Have you tried to use the `split` method?

Reply:
I was hoping not to iterate the whole string. Should I change it into sth like this?

```rust
let parts: Vec<&str> = text.split('/').collect();
let package_name = if text.starts_with('@') && parts.len() > 2 {
    // Scoped package: Include the first two parts (@scope/package)
    parts[..2].join("/")
} else if !text.starts_with('@') && parts.len() > 1 {
    // Unscoped package: Include the first part before the first slash
    parts[0].to_string()
} else {
    // No slashes found, or it does not match the conditions above
    text.to_string()
};
```

Reply:
I don't think you need to do a `.collect`. The `split` returns an iterator, so you can:
- Use `.count` instead of `.len`
- Use `.next()` instead of `parts[0]` and `.next().next()` instead of `parts[1]`. These operations are safer because they return an `Option`

Reply:
@ematipico Wouldn't `count()` consume the iterator so I can't use `next()` later?

Reply:
> Wouldn't count() consume the iterator so I can't use next() later?

It will.
Why do you need to count the number of parts?

Reply:
> @ematipico Wouldn't `count()` consume the iterator so I can't use `next()` later?

Although we can do things like this:
```rs
let part_1 = iter.next();
let part_2 = iter.next();
let len = iter.count();
```

I *think* it should work in our case, but I don't recall what `.next` return from a `Split` iterator.

Reply:
Updated in d6855ae. Thanks for all the suggestions @Conaclos @ematipico, I'm still struggling with rust 😅.

Reply:
Don't worry; any questions are welcome, and we're happy to help. We all have to start from somewhere!